include/annealing/gaussian_annealing.h

```diff
@@ -0,0 +1,215 @@
+        sum = 0.0;
+        for (typename std::vector<FT>::iterator it = dists.begin(); it != dists.end(); ++it) {
+            sum += std::exp(-mid * std::pow(*it, 2)) / (2 * (*it) * std::sqrt(M_PI * mid));
+        }
+
+        if (sum < frac * error) {
+            upper = mid;
+        } else {
+            lower = mid;
+        }
+    }
+
+    a_vals.push_back((upper + lower) / 2.0);
+    error = (1.0 - frac) * error;
+}
+
+
+// Compute a_{i+1} when a_i is given
+template <class T1, typename FT>
+void get_next_gaussian(T1 K,Point &p, FT a, int N, FT ratio, FT C, vars_g var, std::vector<FT> &a_vals){
+
+    FT last_a = a, last_ratio = 0.1;
+    //k is needed for the computation of the next variance a_{i+1} = a_i * (1-1/d)^k
+    FT k = 1.0;
+    const FT tol = 0.00001;
+    bool done=false, print=var.verbose;
+    std::vector<FT> fn(N,0);
+    std::list<Point> randPoints;
+
+    // Set the radius for the ball walk if it is requested
+    if (var.ball_walk) {
+        if (var.delta < 0.0) {
+            var.delta = 4.0 * var.che_rad / std::sqrt(std::max(1.0, last_a) * FT(var.n));
+        }
+    }
+    //sample N points using hit and run or ball walk
+    rand_gaussian_point_generator(K, p, N, var.walk_steps, randPoints, last_a, var);
+
+    typename std::vector<FT>::iterator fnit;
+    while(!done){
+        a = last_a*std::pow(ratio,k);
+
+        fnit = fn.begin();
+        for(std::list<Point>::iterator pit=randPoints.begin(); pit!=randPoints.end(); ++pit, fnit++){
+            *fnit = eval_exp(*pit,a)/eval_exp(*pit, last_a);
+        }
+        std::pair<FT,FT> mv = getMeanVariance(fn);
+
+        // Compute a_{i+1}
+        if(mv.second/(mv.first * mv.first)>=C || mv.first/last_ratio<1.0+tol){
+            if(k!=1.0){
+                k=k/2;
+            }
+            done=true;
+        }else{
+            k=2*k;
+        }
+        last_ratio = mv.first;
+    }
+    a_vals.push_back(last_a*std::pow(ratio, k));
+}
+
+
+// Compute the sequence of spherical gaussians
+template <class T1, typename FT>
+void get_annealing_schedule(T1 K, FT radius, FT ratio, FT C, FT frac, int N, vars_g var, FT &error, std::vector<FT> &a_vals){
+    bool print=var.verbose;
+    // Compute the first gaussian
+    get_first_gaussian(K, radius, frac, var, error, a_vals);
+    if(print) std::cout<<"first gaussian computed\n"<<std::endl;
+
+    FT a_stop = 0.0, curr_fn = 2.0, curr_its = 1.0;
+    const FT tol = 0.001;
+    int it = 0, n = var.n, steps, coord_prev;
+    const int totalSteps= ((int)150/error)+1;
+    std::list<Point> randPoints;
+
+    if(a_vals[0]<a_stop) {
+        a_vals[0] = a_stop;
+    }
+
+    Point p(n);
+
+    if(print) std::cout<<"Computing the sequence of gaussians..\n"<<std::endl;
+
+    Point p_prev=p;
+    std::vector<FT> lamdas(K.num_of_hyperplanes(),NT(0));
+    while (curr_fn/curr_its>(1.0+tol) && a_vals[it]>=a_stop) {
+        // Compute the next gaussian
+        get_next_gaussian(K, p, a_vals[it], N, ratio, C, var, a_vals);
```

Comment:
this could return the last gaussian an then later check if this is last and push it to `a_vals`. Thus avoid to pass `a_vals` as parameter to `get_next_gaussian` and unnecessary `push` `pop` commands

Reply:
done